File: program/match.ts
const url = "/users/123/posts";
const a = 0,
  b = 1,
  c = 2;

const routeData: [number, Array<[string, number]>][] = [];
routeData[1] = [a, []];
routeData[3] = [b, [["id", 1]]];
routeData[4] = [c, [["id", 1]]];
const matchResult = url.match(/^\/(?:$()|users\/([^\/]+)(?:$()|\/posts$()))/);
console.log(matchResult);
if (matchResult) {
  const [handler, paramMap] = routeData[matchResult.indexOf("", 1)];
}

File: program/match.pl
my $url = '/users/123/posts';
my ($a, $b, $c) = (0, 1, 2);

use re 'eval';

my @routes = ($a, $b, $c);

my @c = ();
my $m = undef;

$url =~ m{
  \A/(?:\z(?{$m=0})
  |users/([^/]+)(?:\z(?{$m=1;@c=($1)})
  |/posts\z(?{$m=2;@c=($1)})))
}x;

use Data::Dumper; use Carp; warn Dumper($m, \@c);
